hudi-platform-service/hudi-metaserver/hudi-metaserver-server/src/main/java/org/apache/hudi/metaserver/store/RelationalDBBasedStorage.java

```diff
@@ -0,0 +1,240 @@
+    params.put("oldInstant", oldInstant);
+    params.put("newInstant", newInstant);
+    return timelineDao.updateBySql("updateInstant", params) == 1;
+  }
+
+  @Override
+  public boolean deleteInstant(long tableId, THoodieInstant instant) throws MetaserverStorageException {
+    Map<String, Object> params = new HashMap<>();
+    params.put("tableId", tableId);
+    params.put("ts", instant.getTimestamp());
+    return timelineDao.deleteBySql("deleteInstant", params) == 1;
+  }
+
+  @Override
+  public List<THoodieInstant> scanInstants(long tableId, List<TState> states, int limit) throws MetaserverStorageException {
+    if (isNullOrEmpty(states)) {
+      throw new MetaserverStorageException("State has to be specified when scan instants");
+    }
+    Map<String, Object> params = new HashMap<>();
+    params.put("tableId", tableId);
+    params.put("states", states.stream().mapToInt(TState::getValue).boxed().collect(Collectors.toList()));
+    params.put("limit", limit);
+    List<InstantBean> instantBeans = timelineDao.queryForListBySql("selectInstantsByStates", params);
+    return instantBeans.stream().map(InstantBean::toTHoodieInstant).collect(Collectors.toList());
+  }
+
+  @Override
+  public List<THoodieInstant> scanInstants(long tableId, TState state, int limit) throws MetaserverStorageException {
+    return scanInstants(tableId, Collections.singletonList(state), limit);
+  }
+
+  @Override
+  public boolean instantExists(long tableId, THoodieInstant instant) throws MetaserverStorageException {
+    InstantBean instantBean = new InstantBean(tableId, instant);
+    List<Long> ids = timelineDao.queryForListBySql("selectInstantId", instantBean);
+    validate(ids, instantBean.toString());
+    return !ids.isEmpty();
+  }
+
+  // todo: check correctness
+  @Override
+  public void saveInstantMetadata(long tableId, THoodieInstant instant, byte[] metadata) throws MetaserverStorageException {
+    InstantBean instantBean = new InstantBean(tableId, instant);
+    Map<String, Object> params = new HashMap<>();
+    params.put("instant", instantBean);
+    params.put("metadata", metadata);
+    // todo: array bytes to longblob
+    timelineDao.insertBySql("insertInstantMetadata", params);
+  }
+
+  @Override
+  public boolean deleteInstantMetadata(long tableId, THoodieInstant instant) throws MetaserverStorageException {
+    InstantBean instantBean = new InstantBean(tableId, instant);
+    return timelineDao.deleteBySql("deleteInstantMetadata", instantBean) == 1;
+  }
+
+  @Override
+  public boolean deleteInstantAllMeta(long tableId, String timestamp) throws MetaserverStorageException {
+    Map<String, Object> params = new HashMap<>();
+    params.put("tableId", tableId);
+    params.put("ts", timestamp);
+    return timelineDao.deleteBySql("deleteInstantAllMetadata", params) >= 1;
+  }
+
+  @Override
+  public byte[] getInstantMetadata(long tableId, THoodieInstant instant) throws MetaserverStorageException {
```

Comment:
the other interface is using Option<byte[]>. we should align them for consistency

Reply:
fix